Rework Player Controls

Chris3606 opened the pull request:
# Issue
The current player controls component has a number of shortcomings:
1. It cannot easily handle modifier keys (shift, ctrl, alt, etc)
2. Its motion system does not allow doing anything with the direction generated other than "attempt to move the Parent via setting its Position value".  This fails to cover some common use cases, including easily implementing the commonly used "bump-attack" in roguelikes.
3. The component is, in at least one environment, duplicating keypresses in some cases (known bug).
4. The amount of time it takes to execute per frame is directly proportional to the number of keybindings added.

# Implemented Solution
This PR attempts to implement a fix for problems 1, 2, and 4 above (cause of 3 is still unknown, have been unable to reproduce in an isolated environment).  It rewrites significant portions of the old system, including the following changes:
- Splits key-handling functionality into its own namespace and renames some things
    - Some of the renames can be reverted if they are not well-received; it is just an attempt at producing a logical organization
- Motion system now calls a motion handling action, passing the generated direction to it.  The motion handler can be customized to produce custom behavior.
- Keybindings may now be specified via InputKey structs (to allow modifier support)
- Class now supports keybindings using modifier keys (shift, alt, etc)
     - Modifier keys may be handled 2 ways; exact-match mode, and non-exact match; differences and purpose are described in class documentation
     - Modifiers include left- and right- variants of each supported modifier key, as well as an "either one" option
     - Support for this involved notable changes to the internals of the class and how it caches input
- Keybindings are now detected/dispatched by iterating over SadConsole's list of pressed keys, since the number of keys pressed should almost always be small relative to the number of keybindings registered.
- Changed motion/action add functions to `SetMotion` and `SetAction`; they now overwrite existing key bindings to the same key.  This should prevent the rather unpredictable behavior that can result from having multiple bindings for the same key at once.
- Changed the way default motions are handled
    - Now handled via a combination of static arrays that define common control schemes, and the new `SetMotions` function
    - See `Player.cs` for an example usage
- Changed ExampleGame to use MonoGame host; this is a miscellaneous change
    - Not strictly required but SFML currently requires additional steps to run on Linux, so ideally to make the example easily runnable for the maximum number of users, the example would use MonoGame
    - There are also MonoGame-specific components such as the FPS counter component that may prove useful in testing

## Notable Things NOT Implemented 
- Bugfix for issue 3; cause still unknown; need to determine how to reliably reproduce
- CAPSLOCK support (making CAPSLOCK optionally equivalent to holding shift); it's possible but could be a future PR and would IMO be a niche future; more concerned with getting a releasable architecture initially

fcheadle commented on TheSadRogue.Integration/Components/Keybindings/InputKey.cs:
```diff
@@ -0,0 +1,192 @@
+            Alt = KeyModifiers.None;
+            foreach (var modifier in modifiers)
+            {
+                switch (modifier)
```
Or a property setter if you prefer:
```CSharp
private readonly KeyModifiers _alt;
public KeyModifiers Alt
{
    get => _alt;
    private set 
    {
            if (_alt != KeyModifiers.None)
                throw new ArgumentException("Only one Alt modifier is allowed.", nameof(modifiers));
            _alt = modifier;
    }
}
```

Chris3606 replied:
It's a bit of a personal preference IMO thing but I'll see what I can do.  I kind of view it as 6-half dozen, there is a little logic in a switch but putting the logic in a function would promote reuse and that's bad in this case because this is only ever something that should happen in init.

fcheadle replied:
Fair enough.

Chris3606 commented on TheSadRogue.Integration/Components/Keybindings/InputKey.cs:
```diff
@@ -0,0 +1,192 @@
+        private static bool MatchesState(Keyboard keyboard, KeyModifiers state, Keys keyLeft, Keys keyRight, KeyModifiers left,
+                                              KeyModifiers right)
+        {
+            if (state == KeyModifiers.None) return true;
```
> If the state == none, we return true whether shift is held or not?
No, this would be the exact match feature.

> Will this not cause some confusion if a user is using non-exact matches, and they have a command for s and shift+s?
It should not, because the items are inserted in to the lists such that items with more modifiers are always first, and as such are always examined first.  If the user has S and Shift-S, Shift-S would have always been checked for first; so this line when checking for S would only occur if we already know that Shift-S won't match.

fcheadle commented:
Last minor nitpick: I think that having the inputs in the `Components` namespace is slightly murky. `InputKeys` and `KeyModifiers`, for example, don't belong in a namespace under `Components`.

Chris3606 commented:
> I think that having the inputs in the Components namespace is slightly murky

I agree; I don't like the current arrangement either.  I wasn't sure of the best way to organize it; perhaps this?

```
- SadRogue.Integration.Keybindings
    - InputKey
    - KeyModifiers
- SadRogue.Components
    - PlayerKeybindingsComponent
```

Any better?  Other ideas?

fcheadle commented:
I would put the `PlayerKeyBindingsComponent` in the `SadRogue.Integration.KeyBindings` namespace, unless you're planning on adding more components that use the keybindings system

Chris3606 commented:
> I would put the PlayerKeyBindingsComponent in the SadRogue.Integration.KeyBindings namespace

I was thinking about doing this too; one thing of note is that SadConsole has its SadConsole.Components namespace which contains components that interact with various systems (including `SurfaceComponentFollowTarget` and some others), so it would be a little less consistent with that.  However, SadConsole also has `EntityRenderer` which is in the `SadConsole.Entities` namespace, but is a component; so it appears SadConsole follows no particular convention in this light so it might be fine.

Chris3606 commented:
Above is now implemented in branch's current state.

Chris3606 commented:
NOTE: PR has been moved from draft status to ready for review.  Changes from the last review:
- Implemented re-organization of namespaces as per above conversation
- Removed `AddMotion` calls in the constructor, so there are no motions added by default
- Provided functions that can add multiple motions and actions respectively at once for convenience
- Provided a number of static arrays in the class that define common movement schemes (ie. arrow keys, numpad, etc).
    - This plus the functions that take set multiple motions at a time constitutes a replacement for default motions; see `Player.cs`)
- Changed `AddMotion` and `AddAction` functions to `SetMotion` and `SetAction`
    - They now override any existing item in their respective lists that is bound to the same key; thus preventing duplicate key bindings from occuring

Thraka reviewed (approved):
Look like a great design!
